.pre-commit-config.yaml

```diff
@@ -438,6 +438,11 @@ repos:
         files: ^airflow/www/.*\.(css|scss|sass)$
         # Keep dependency versions in sync w/ airflow/www/package.json
         additional_dependencies: ['stylelint@13.3.1', 'stylelint-config-standard@20.0.0']
+      - id: version-sync
+        name: Version sync
+        files: ^airflow/version.py$|setup.py
+        entry: ./scripts/ci/pre_commit/pre_commit_sync_version.sh
```

Comment:
What's this needed for?

Reply:
I was just about to write about it. The previous way of retrieving version gave already hard time in previous scenarios when setup.py was used while airflow was already installed and in pre-commits. It sometimes did not found version.py. Therefore I've added a change to synchronize version in setup.py and version.py at the pre-commit level. This way there is no importlib call during setup.py execution - version is hard-coded, so there is no risk any problems will happen.

Reply:
This way @ashb -> we do not have any dynamic importing while building/preparing the packages. IMHO it's much more robust and healthy if the version is hard-coded. It is especially a problem in case we want to import stuff like "EXTRAS" etc in setup.py and run those check in various "environments" - for example in pre-commits and in case of scripts that are running outside of "airflow" folder, or when we have airflow installed as pypi package rather than fro sources.

Reply:
Not thrilled about this change.

When, and why, are we importing the top-level setup.py when Airflow is already installed?

Reply:
(This doesn't block us merging this for now)

Reply:
There are at lest few cases:

* We are importing it in some python pre-commits to verify setup.py consistency, ordering, completeness 

* We are imporint it when  automatically generating provider package setup.py's or when we generate README files for those (we are retrieving information about dependencies for each package so that we can produce a documentation showing those dependencies and add the dependencies in the generated setup.py for providers using a single source of truth.

Reply:
Happy if you have other proposals for those.

The provider setup.py and README generation is difficult to be done otherwise unless we built in some dump of dependencies from setup.py in some structured way, but we would need to do it as pre-commit whenever setup.py changes.

We already do that for another case - to generate cross-provider dependencies https://github.com/apache/airflow/blob/master/airflow/providers/dependencies.json - this file is a structured dump of which provider depends on which. It is done from the sources rather than from setup.py automatically using pre-commit.

But I think it would not be a good way to keep that information in two places (setup.py and another .json or similar).

Happy to discuss what other approaches we might take there if you can propose something.

Reply:
For inspiration - this is how we generate those depenencies.json . I think they are useful to keep to see the dependencies: 
```
      - id: build-providers-dependencies
        name: Build cross-dependencies for providers packages
        entry: "./scripts/ci/pre_commit/pre_commit_build_providers_dependencies.sh"
        language: system
        files: ^airflow/providers/.*\.py$|^tests/providers/.*\.py$
        pass_filenames: false
```